pkg/ibm_plugin/server/plugin.go

```diff
@@ -335,127 +336,198 @@ func (s *IBMPluginServer) AddPermitListRules(ctx context.Context, req *paraglide
+	vpcAddressSpaces, err := cloudClient.GetVpcCIDR(*requestVPCData.ID)
 	if err != nil {
 		return nil, err
 	}
-	defer conn.Close()
-	client := paragliderpb.NewControllerClient(conn)
-
-	gwID := "" // global transit gateway ID for vpc-peering.
-	for _, ibmRule := range ibmRulesToAdd {
 
-		// TODO @cohen-j-omer Connect clouds if needed:
-		// 1. use the orchestratorClient's GetUsedAddressSpaces to get used addresses.
-		// 2. if the rule's remote address resides in one of the clouds create a vpn gateway.
+	// get current rules in SG and record their hash values
+	sgRules, err := cloudClient.GetSecurityRulesOfSG(requestSGID)
+	if err != nil {
+		return nil, err
+	}
 
-		// get the VPCs and clients to search if the remote IP resides in any of them
-		clients, err := s.getAllClientsForVPCs(cloudClient, rInfo.ResourceGroup, false)
+	// Get used address spaces of all clouds
+	controllerConn, err := grpc.Dial(s.orchestratorServerAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, fmt.Errorf("unable to establish connection with orchestrator: %w", err)
+	}
+	defer controllerConn.Close()
+	controllerClient := paragliderpb.NewControllerClient(controllerConn)
+	addressSpaceMappings, err := controllerClient.GetUsedAddressSpaces(context.Background(), &emptypb.Empty{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to get used address spaces: %w", err)
+	}
+	gwID := "" // global transit gateway ID for vpc-peering.
+	for _, paragliderRule := range req.Rules {
+		// translate paraglider rule to IBM rules to compare hash values with current rules.
+		// multiple ibm rules can be returned due to multiple possible targets
+		ibmRules, err := sdk.ParagliderToIBMRule(requestSGID, paragliderRule)
 		if err != nil {
 			utils.Log.Printf("Failed to get remote vpc: %v.\n", err)
 			return nil, err
 		}
-		remoteVPC := ""
-		for vpcID, client := range clients {
-			if isRemoteInVPC, _ := client.IsRemoteInVPC(vpcID, ibmRule.Remote); isRemoteInVPC {
-				remoteVPC = vpcID
-				break
-			}
+
+		// peeringCloudInfos contains cloud info associated with the specified rule's targets
+		peeringCloudInfos, err := utils.GetPermitListRulePeeringCloudInfo(paragliderRule, addressSpaceMappings.AddressSpaceMappings)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get peering cloud infos: %w", err)
 		}
-		// if the remote resides inside a paraglider VPC that isn't the request resource's VPC, connect them
-		if remoteVPC != "" && remoteVPC != *requestVPCData.ID {
-			utils.Log.Printf("The following rule's remote is targeting a different IBM VPC\nRule: %+v\nVPC:%+v", ibmRule, remoteVPC)
-			// fetch or create transit gateway
-			if len(gwID) == 0 { // lookup optimization, use the already fetched gateway ID if possible
-				gwID, err = cloudClient.GetOrCreateTransitGateway(region)
+		// connect clouds if needed
+		for i, peeringCloudInfo := range peeringCloudInfos {
+			if peeringCloudInfo == nil { // public IP address
```

Comment:
```suggestion
			if peeringCloudInfo == nil {
```
It may be nil for other reasons. Let's keep that info in one place in the `GetPermitListRulePeeringCloudInfo` docstring.